stackslib/src/net/unsolicited.rs

```diff
@@ -0,0 +1,1030 @@
+                sortdb,
+                consensus_hash,
+            )
+        } else {
+            inv.set_block_available(
+                &self.burnchain,
+                outbound_neighbor_key,
+                sortdb,
+                consensus_hash,
+            )
+        };
+
+        let block_sortition_height = match res {
+            Ok(Some(block_height)) => block_height,
+            Ok(None) => {
+                debug!(
+                    "{:?}: We already know the inventory state in {} for {}",
+                    &self.local_peer, outbound_neighbor_key, consensus_hash
+                );
+                return Ok(None);
+            }
+            Err(NetError::NotFoundError) => {
+                // is this remote node simply ahead of us?
+                if let Some(convo) = self.peers.get(&event_id) {
+                    if self.chain_view.burn_block_height < convo.burnchain_tip_height {
+                        debug!("{:?}: Unrecognized consensus hash {}; it is possible that {} is ahead of us", &self.local_peer, consensus_hash, outbound_neighbor_key);
+                        return Err(NetError::NotFoundError);
+                    }
+                }
+                // not ahead of us -- it's a bad consensus hash
+                debug!("{:?}: Unrecognized consensus hash {}; assuming that {} has a different chain view", &self.local_peer, consensus_hash, outbound_neighbor_key);
+                return Ok(None);
+            }
+            Err(NetError::InvalidMessage) => {
+                // punish this peer
+                info!(
+                    "Peer {:?} sent an invalid update for {}",
+                    &outbound_neighbor_key,
+                    if microblocks {
+                        "streamed microblocks"
+                    } else {
+                        "blocks"
+                    }
+                );
+                self.bans.insert(event_id);
+
+                if let Some(outbound_event_id) = self.events.get(&outbound_neighbor_key) {
+                    self.bans.insert(*outbound_event_id);
+                }
+                return Ok(None);
+            }
+            Err(e) => {
+                warn!(
+                    "Failed to update inv state for {:?}: {:?}",
+                    &outbound_neighbor_key, &e
+                );
+                return Ok(None);
+            }
+        };
+        Ok(Some(block_sortition_height))
+    }
+
+    /// Buffer a message for re-processing once the burnchain view updates.
+    /// If there is no space for the message, then silently drop it.
+    fn buffer_data_message(&mut self, event_id: usize, msg: StacksMessage) {
+        let Some(msgs) = self.pending_messages.get_mut(&event_id) else {
+            self.pending_messages.insert(event_id, vec![msg]);
+            debug!(
+                "{:?}: Event {} has 1 messages buffered",
+                &self.local_peer, event_id
+            );
+            return;
+        };
```

Comment:
Avoid repeated hashmap lookup with entry:
```suggestion
        let msgs = match self.pending_messages.entry(&event_id) {
            Entry::Occupied(ref mut x) => x.get_mut(),
            Entry::Vacant(ref mut slot) => {
                    slot.insert(vec![msg]);
                    debug!(
                        "{:?}: Event {} has 1 messages buffered",
                        &self.local_peer, event_id
                    );
                   return;
        };
```

Reply:
Siphash is extremely fast (especially on a machine word sized key), and I personally avoid the Entries API because it adds needless cognitive overhead to the reader.